### PR TITLE
HPCC-25187 Minimize memory consumption by reducing space allocated to arrays

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith

File: ecl/hql/hqlexpr.cpp [hunk 1]
```diff
@@ -6009,7 +6011,7 @@ void CHqlSelectBaseExpression::setOperands(IHqlExpression * left, IHqlExpression
 {
     //Need to be very careful about the order that this is done in, since queryType() depends on operand2
     unsigned max = attr ? 3 : 2;
-    operands.ensure(max);
+    operands.ensureSpace(max);
```
Comment (jakesmith) [inline ecl/hql/hqlexpr.cpp:6014]: shouldn't this be a ensureCapacity(max) ?

File: system/jlib/jarray.cpp [hunk 2]
```diff
@@ -46,15 +46,10 @@ void Allocator::_space(size32_t iSize)
...
 }
 
-void Allocator::_ensure(aindex_t next, size32_t iSize)
+void Allocator::_ensure(aindex_t req, size32_t iSize)
 {
-    aindex_t req = used + next;
     if (req > max )
-    {
-        if ((max == 0) && (req < DOUBLE_LIMIT))
-            max = ensurePowerOfTwo(req);
```
Comment (jakesmith) [inline system/jlib/jarray.cpp:55]: removed use of ensurePowerOfTwo.
ensurePowerOfTwo definition still exists, but not used.

File: system/jlib/jarray.cpp [hunk 3]
```diff
@@ -103,6 +98,26 @@ void Allocator::_reallocate(aindex_t newLen, size32_t itemSize)
...
+        return;
+
+    if (used > newLen)
+        throw MakeStringException(0, "Reallocate an array would contain too few items: itemSize = %d, trying to allocate %d items contains %u",itemSize, newLen, used);
+
+    size_t allocSize = (size_t)itemSize * newLen;
+    void *newhead = realloc(_head, allocSize);
+    if (!newhead)
+    {
+        IERRLOG("Out of memory in Array allocator: itemSize = %d, trying to allocate %d items", itemSize, max);
```
Comment (jakesmith) [inline system/jlib/jarray.cpp:113]: trivial: should be %u's ?

File: system/jlib/jarray.cpp [hunk 4]
```diff
@@ -103,6 +98,26 @@ void Allocator::_reallocate(aindex_t newLen, size32_t itemSize)
...
+        throw MakeStringException(0, "Out of memory in Array allocator: itemSize = %d, trying to allocate %d items",itemSize, max);
```
Comment (jakesmith) [inline system/jlib/jarray.cpp:114]: makeStringExceptionV and should be %u's ?

**Review by jakesmith — COMMENTED** [review]
@ghalliday - a couple of comments/questions.

File: ecl/hql/hqlexpr.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlexpr.cpp:6014]: They will have an identical effect, it doesn't really matter..

File: system/jlib/jarray.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jarray.cpp:55]: deleted

File: system/jlib/jarray.cpp (hunk 3 above)
Comment (ghalliday) [reply system/jlib/jarray.cpp:113]: It was copied code.  I will change them all.

File: system/jlib/jarray.cpp (hunk 4 above)
Comment (ghalliday) [reply system/jlib/jarray.cpp:114]: Leaving as it is - the code was cloned.

**Comment by ghalliday** [comment]
@richardkchapman I responded to most of Jake's comments, and squashed the change.  Please review/merge.

**Review by richardkchapman — APPROVED** [review]

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Failing some tests

**Comment by AttilaVamos** [comment]
@ghalliday I am sure that Hthor schedule1 is not related to the changes in this PR. Re-run Smoketest to get a clean result.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**